Simplenote/src/main/java/com/automattic/simplenote/viewmodels/TagDialogViewModel.kt

```diff
@@ -0,0 +1,98 @@
+package com.automattic.simplenote.viewmodels
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import com.automattic.simplenote.R
+import com.automattic.simplenote.analytics.AnalyticsTracker
+import com.automattic.simplenote.models.Tag
+import com.automattic.simplenote.repositories.TagsRepository
+
+class TagDialogViewModel(private val tagsRepository: TagsRepository) : ViewModel() {
+    private val _uiState = MutableLiveData<UiState>()
+    val uiState: LiveData<UiState> = _uiState
+
+    private val _event = SingleLiveEvent<TagDialogEvent>()
+    val event: LiveData<TagDialogEvent> = _event
+
+    fun start(tag: Tag) {
+        _uiState.value = UiState(tag.name, tag)
+    }
+
+    fun close() {
+        _event.postValue(CloseEvent)
+    }
+
+    fun updateUiState(tagName: String) {
+        if (tagName.isEmpty()) {
+            _uiState.value = _uiState.value!!.copy(tagName = tagName, errorMsg = R.string.tag_error_empty)
+            return
+        }
+
+        if (tagName.contains(" ")) {
+            _uiState.value = _uiState.value!!.copy(tagName = tagName, errorMsg = R.string.tag_error_spaces)
+            return
+        }
+
+        if (!tagsRepository.isTagValid(tagName)) {
+            _uiState.value = _uiState.value!!.copy(tagName = tagName, errorMsg = R.string.tag_error_length)
+            return
+        }
+
+        _uiState.value = _uiState.value!!.copy(tagName = tagName, errorMsg = null)
+    }
+
+    fun renameTagIfValid() {
+        val currentState = uiState.value
+
+        if (currentState == null) {
+            _event.postValue(ShowErrorEvent)
+            return
+        }
+
+        // If the tag did not changed, do not anything
+        if (currentState.tagName == currentState.oldTag.name) {
+            _event.postValue(FinishEvent)
+            return
+        }
+
+        // If there is another tag with the same name
+        if (!tagsRepository.isTagMissing(currentState.tagName)) {
```

Comment:
Yes, it does not cover all the behavior. `isTagMissing` covers `hasCanonicalTag` since is the same method as `hasCanonicalOfLexical` (just opposite answers), but does not cover `!isRenamingToLexicalTag`. `!isRenamingToLexicalTag` checks for the case in which the tag is not identical the current one but the lexical tag is the same. This would show the merge dialog unnecessarily. Fixed in 7bf0745b86e4630215e26b79498bc0b90ba0f196 (added `isTagConflict` to `TagsRepository`) and bbfc4c1f7e8385de71a4f9e08f6655d173824141 (used `isTagConflict` in ViewModel and fixed tests).

Reply:
> isTagMissing covers hasCanonicalTag since is the same method as hasCanonicalOfLexical (just opposite answers).

This is true!

> ...but does not cover !isRenamingToLexicalTag. !isRenamingToLexicalTag checks for the case in which the tag is not identical the current one but the lexical tag is the same. This would show the merge dialog unnecessarily.

Nice finding!

> Fixed in 7bf0745 (added isTagConflict to TagsRepository) and bbfc4c1 (used isTagConflict in ViewModel and fixed tests).

🚀

Reply:
PS: I think missing this behavior changes is very easy when not doing these kind of refactoring, migration, modernization steps commit-by-commit, guarding those commits with tests, even TDD wise if possible. Thus, you are seeing me pushing towards this direction a bit too enthusiastically. 😅

Reply:
👋 @dandominguez04 !

> isTagMissing covers hasCanonicalTag since is the same method as hasCanonicalOfLexical (just opposite answers).

About that, I still wonder why the initial author choose to use the `hasCanonicalOfLexical(...)` instead of `isTagMissing(...)`. Do you think the `isTagMissing(...)` is better suited here, thus you choose to change it or?

Reply:
HI @ParaskP7 !

> PS: I think missing this behavior changes is very easy when not doing these kind of refactoring, migration, modernization steps commit-by-commit, guarding those commits with tests, even TDD wise if possible. Thus, you are seeing me pushing towards this direction a bit too enthusiastically.

Yes, I think the UI tests are missing some tests regarding encoding of tags and the canonical form of the tags. 

> About that, I still wonder why the initial author choose to use the hasCanonicalOfLexical(...) instead of isTagMissing(...). Do you think the isTagMissing(...) is better suited here, thus you choose to change it or?

Maybe the author chose that method to be more explicit on naming about the canonical form of the tag. I chose `isTagMissing(...)` because I did not want to blow the Repository with multiple methods that have the same behavior. I think `isTagMissing(...)` does the job here because we are looking to see if there another tag. The same method is used for validation when adding a new tag, thus, we have uniform behavior across screens.

Reply:
👋 @dandominguez04 !

Thanks for your responses.

> Yes, I think the UI tests are missing some tests regarding encoding of tags and the canonical form of the tags.

True, but I also think that this might be tricky to test anyway.

> Maybe the author chose that method to be more explicit on naming about the canonical form of the tag. I chose isTagMissing(...) because I did not want to blow the Repository with multiple methods that have the same behavior. I think isTagMissing(...) does the job here because we are looking to see if there another tag. The same method is used for validation when adding a new tag, thus, we have uniform behavior across screens.

I wouldn't worry about blowing the `Repository`, it is anyway not that big atm, especially if the right naming is used. However, you other arguments might suffice for now, so I am okay to keeping it as is for now. In the future, we might merge those two similar utility methods into a unified one, with a more appropriate naming, or choose to rollback to using the original utility method.

But, for now we are okay, thanks for taking the time to think about it. 🙏